crates/sui-kvstore/src/client.rs

```diff
@@ -0,0 +1,176 @@
+        values
+            .into_iter()
+            .map(|(k, v)| bcs::to_bytes(v.borrow()).map(|v| (k, v)))
+            .collect::<Result<_, _>>()
+    }
+}
+
+pub struct DynamoDbClient {
+    client: dynamodb::Client,
+    namespace: Option<String>,
+}
+
+impl DynamoDbClient {
+    pub async fn new(config: &KVStoreConfig) -> Self {
+        let credentials = Credentials::new(
+            &config.aws_access_key_id,
+            &config.aws_secret_access_key,
+            None,
+            None,
+            "dynamodb",
+        );
+        let aws_config = aws_config::from_env()
+            .credentials_provider(credentials)
+            .region(Region::new(config.aws_region.clone()))
+            .load()
+            .await;
+        let client = dynamodb::Client::new(&aws_config);
+        Self {
+            client,
+            namespace: config.namespace.clone(),
+        }
+    }
+
+    fn table_name(&self, table: KVTable) -> String {
+        match &self.namespace {
+            Some(suffix) => format!("{}-{}", table, suffix),
+            None => table.to_string(),
+        }
+        .to_lowercase()
+    }
+}
+
+#[async_trait]
+impl KVWriteClient for DynamoDbClient {
+    async fn multi_set<V: Serialize>(
+        &mut self,
+        table: KVTable,
+        values: impl IntoIterator<Item = (String, V)> + std::marker::Send,
+    ) -> anyhow::Result<()> {
+        let items = Self::serialize_values(values)?
+            .into_iter()
+            .map(|(digest, bytes)| {
+                WriteRequest::builder()
+                    .set_put_request(Some(
+                        PutRequest::builder()
+                            .item("digest", AttributeValue::S(digest))
+                            .item("bytes", AttributeValue::B(Blob::new(bytes)))
+                            .build(),
+                    ))
+                    .build()
+            })
+            .collect();
+        self.client
+            .batch_write_item()
+            .set_request_items(Some(HashMap::from([(self.table_name(table), items)])))
+            .send()
+            .await?;
+        Ok(())
+    }
+
+    async fn get_state(&self, state: State) -> anyhow::Result<Option<u64>> {
+        let item = self
+            .client
+            .get_item()
+            .table_name(self.table_name(KVTable::State))
+            .key("id", AttributeValue::S(state.to_string()))
+            .send()
+            .await?;
+        if let Some(output) = item.item() {
+            if let AttributeValue::N(progress) = &output["value"] {
+                return Ok(Some(progress.parse()?));
+            }
+        }
+        Ok(None)
+    }
+
+    async fn update_state(&mut self, state: State, value: u64) -> anyhow::Result<()> {
+        self.client
+            .put_item()
+            .table_name(self.table_name(KVTable::State))
+            .item("id", AttributeValue::S(state.to_string()))
+            .item("value", AttributeValue::N(value.to_string()))
```

Comment:
should we use a conditional operator here to make sure that the watermark doesn't go backwards (which might happen if someone starts up a second writer)

Reply:
idk, on the other hand we might have a reason in the future to backfill the table the second time (adding attributes for GSIs and whatnot)

Reply:
i think it would be cleaner to use a different state key for each unique backfill process we might run later.